### PR TITLE
add KeyError for key and feature type

### LINKED ISSUE #766: [EN] Add warnings for key types and feature types (since they are different)
### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### Feathr version

0.8

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 20.0)**:
- **Python version**:
- **Spark version, if reporting runtime issue**:

### Describe the problem

Currently the Feathr key type and the feature type is different and users usually mix those (we've seen it multiple times). So we should:

1. Add type checks for the type of key (ValueType.INT32) and feature types (from feathr import INT32)
2. give warnings out if users mix those

### Tracking information

_No response_

### Code to reproduce bug

_No response_

### What component(s) does this bug affect?

- [ ] `Python Client`: This is the client users use to interact with most of our API. Mostly written in Python.
- [ ] `Computation Engine`: The computation engine that execute the actual feature join and generation work. Mostly in Scala and Spark.
- [ ] `Feature Registry API`: The frontend API layer supports SQL, Purview(Atlas) as storage. The API layer is in Python(FAST API)
- [ ] `Feature Registry Web UI`: The Web UI for feature registry. Written in React

### PR DESCRIPTION
**Pull request by Yuqing-cat**
Signed-off-by: Yuqing Wei <weiyuqing021@outlook.com>

## Description

Resolves #766 

## How was this PR tested?

Test `test_dtype.py` added under unit folder with 2 functions:
- test_key_type
- test_feature_type

image

## Does this PR introduce any user-facing changes?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure to clarify your proposed changes.

User will get KeyError if the type of Key or Feature is not correct.

### CODE REVIEW

**Review by xiaoyongzhu — APPROVED** [review]

**Comment by xiaoyongzhu** [comment]
The test failure should be unrelated IMO

File: feathr_project/feathr/definition/feature.py [hunk 1]
```diff
@@ -30,6 +30,11 @@ def __init__(self,
                  registry_tags: Optional[Dict[str, str]] = None,
                  ):
         FeatureBase.validate_feature_name(name)
+
+        # Validate the feature type
+        if not isinstance(feature_type, FeatureType):
+            raise KeyError(f'Feature type must be a FeatureType class, like INT32, but got {feature_type}')
```
Comment (blrchen) [inline feathr_project/feathr/definition/feature.py:36]: Suggest change to {feature_type} is not a valid feature type, please use one of following valid values: blabla (dump all enum values)

Comment (Yuqing-cat) [reply feathr_project/feathr/definition/feature.py:36]: It's feasible for `KeyType`, which is a `enum`, but cannot be applied to `FeatureType` as it's a base class. 

From my perspective, message looks too long to dump all enum values: 

KeyType: (all of them needs to add `ValueType` as prefix)
    UNSPECIFIED = 0
    BOOL = 1
    INT32 = 2
    INT64 = 3
    FLOAT = 4
    DOUBLE = 5
    STRING = 6
    BYTES = 7

FeatureType:
BOOLEAN = BooleanFeatureType()
INT32 = Int32FeatureType()
INT64 = Int64FeatureType()
FLOAT = FloatFeatureType()
DOUBLE = DoubleFeatureType()
STRING = StringFeatureType()
BYTES = BytesFeatureType()
FLOAT_VECTOR = FloatVectorFeatureType()
INT32_VECTOR = Int32VectorFeatureType()
INT64_VECTOR = Int64VectorFeatureType()
DOUBLE_VECTOR = DoubleVectorFeatureType()

Comment (windoze) [reply feathr_project/feathr/definition/feature.py:36]: Maybe we should use different sets of names for different types, but that'll be a breaking change.

Comment (xiaoyongzhu) [reply feathr_project/feathr/definition/feature.py:36]: > Maybe we should use different sets of names for different types, but that'll be a breaking change.

Agree, that might be ideal but will introduce breaking change. I think we can simply using this check to help users know what's wrong there

**Pull request merged**